### ISSUE TITLE
td.replace throws with objects created with Object.create

### ISSUE THREAD

**Issue by mcous** [testdouble/testdouble.js#257]
First off, I'm a big fan of the project, so thank you!

Second off, I'm trying to mock a module using `td.replace` that exports an object created with `Object.create`, but `td.replace` throws. I haven't dug into testdouble's code yet, but my gut says it has something to do with enumerability (because the function properties of the object are non-enumerable; maybe related to #119?). Is this expected behavior?

Reproduce with:

``` js
// index.js
const td = require('testdouble')

try {
  const obj = td.replace('./normal-object')

  console.log('Normal object')
  console.log(obj)
  console.log(obj.foo)
  console.log(obj.bar)
} catch (e) {
  console.log('Failed to replace normal object')
  console.error(e)
}

console.log()

try {
  const obj = td.replace('./non-enumerable-object')

  console.log('Object with non-enumerable properties')
  console.log(obj)
  console.log(obj.foo)
  console.log(obj.bar)
} catch (e) {
  console.error('Failed to replace object with non-enumerable properties')
  console.error(e)
}

console.log()

try {
  const obj = td.replace('./object-from-create')

  console.log('Object from Object.create')
  console.log(obj)
  console.log(obj.foo)
  console.log(obj.bar)
} catch (e) {
  console.error('Failed to replace object from Object.create')
  console.error(e)
}
```

``` js
// normal-object.js
module.exports = {
  foo: () => 'foo',
  bar: () => 'bar'
}

```

``` js
// non-enumerable-object.js
module.exports = Object.defineProperties({}, {
  foo: {
    enumerable: false,
    value: () => 'foo'
  },

  bar: {
    enumerable: false,
    value: () => 'bar'
  }
})

```

``` js
// object-from-create.js
module.exports = Object.create({
  foo: () => 'foo',
  bar: () => 'bar'
})
```

Output

``` shell
$ node index.js
Normal object
{ foo: { [Function: testDouble] toString: [Function] },
  bar: { [Function: testDouble] toString: [Function] },
  toString: [Function] }
{ [Function: testDouble] toString: [Function] }
{ [Function: testDouble] toString: [Function] }

Object with non-enumerable properties
{ toString: [Function] }
[Function: value]
[Function: value]

Failed to replace object from Object.create
Error: Error: testdouble.js - td.replace - "./object-from-create" property was found, but test double only knows how to replace functions, constructors, & objects containing functions (its value was "[object Object]").
    at Object.error (/Users/mc/sandbox/td-object-create/node_modules/testdouble/lib/log.js:23:13)
    at exports.default (/Users/mc/sandbox/td-object-create/node_modules/testdouble/lib/replace/imitate.js:45:26)
    at exports.default (/Users/mc/sandbox/td-object-create/node_modules/testdouble/lib/replace/module.js:12:41)
    at exports.default [as replace] (/Users/mc/sandbox/td-object-create/node_modules/testdouble/lib/replace/index.js:9:29)
    at Object.<anonymous> (/Users/mc/sandbox/td-object-create/index.js:34:18)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
```

**Comment by searls** [comment]
Version?

**Comment by mcous** [comment]
Sorry, didn't mean to leave out the details!

td: v3.0.0
node: v6.10.2
npm: v5.0.2

**Comment by searls** [comment]
This clause is the offending line.

https://github.com/testdouble/testdouble.js/blob/master/src/replace/imitate.js#L12

I wonder if there is a lodash method for what we're trying to suss-out here?

**Comment by mcous** [comment]
There's _.isObjectLike, but that returns `true` for arrays, so it may not be exactly right. It looks like https://github.com/testdouble/testdouble.js/blob/master/src/object.js#L16 might be a problem, too once we're past that

**Comment by searls** [comment]
Another hiccup: `_.functions()` returns an empty array for objects created with `Object.create()`

**Comment by searls** [comment]
Note that this is also a dupe of #224, but I'm spending more time taking it at face value than I did then

**Comment by searls** [comment]
You could use `_.functionsIn` for the above, FWIW (thanks @jasonkarns). The broader issue here, though, is that in some places we grab non-enumerable properties (see `src/util/copy-properties`) and in other areas we pull in inherited properties (see `src/util/get-all-custom-prototypal-function-names.js`), all depending on whether a user happened to pass a thing to `td.func`, `td.constructor`, or `td.object` and this inconsistency is rightly driving me up a wall.

I think any change we make here should start with a "one property-finding functions" to rule them all, which does all of:
* Includes enumerable and non-enumerable properties (e.g. `Object.getOwnPropertyNames`)
* Includes inherited properties from custom types
* Excludes inhereited properties from built-in types (e.g. Object.prototype.toString)

From that, a second module could filter down just those that properties whose values pass an `_.isFunction` check, and that's what I'd want to use here and anywhere else we're using `_.functions()`

**Comment by mcous** [comment]
> this inconsistency is rightly driving me up a wall

😄 

But yeah, makes total sense. Should #262 also be considered for this one function to rule them all? Also, happy to PR / help in whatever way possible

**Comment by searls** [comment]
Yes, I view this as a prerequisite to doing #262.

**Comment by searls** [comment]
Ok, the big deep property clone/replace/imitation feature shipped in 3.2.0
